chain/store/splitstore/splitstore.go

```diff
@@ -0,0 +1,414 @@
+package splitstore
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/bmatsuo/lmdb-go/lmdb"
+
+	blocks "github.com/ipfs/go-block-format"
+	cid "github.com/ipfs/go-cid"
+	dstore "github.com/ipfs/go-datastore"
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/lotus/chain/store"
+	"github.com/filecoin-project/lotus/chain/types"
+	bstore "github.com/filecoin-project/lotus/lib/blockstore"
+)
+
+const CompactionThreshold = 5 * build.Finality
+
+var baseEpochKey = dstore.NewKey("baseEpoch")
+
+var log = logging.Logger("splitstore")
+
+type SplitStore struct {
+	baseEpoch abi.ChainEpoch
+	curTs     *types.TipSet
+
+	cs *store.ChainStore
+	ds dstore.Datastore
+
+	hot  bstore.Blockstore
+	cold bstore.Blockstore
+
+	snoop TrackingStore
+
+	stateMx    sync.Mutex
+	compacting bool
+
+	env *lmdb.Env
+}
+
+var _ bstore.Blockstore = (*SplitStore)(nil)
+
+// Blockstore interface
+func (s *SplitStore) DeleteBlock(cid cid.Cid) error {
+	// afaict we don't seem to be using this method, so it's not implemented
+	return errors.New("DeleteBlock not implemented on SplitStore; don't do this Luke!") //nolint
+}
+
+func (s *SplitStore) Has(cid cid.Cid) (bool, error) {
+	has, err := s.hot.Has(cid)
+
+	if err != nil {
+		return false, err
+	}
+
+	if has {
+		return true, nil
+	}
+
+	return s.cold.Has(cid)
+}
+
+func (s *SplitStore) Get(cid cid.Cid) (blocks.Block, error) {
+	blk, err := s.hot.Get(cid)
+
+	switch err {
+	case nil:
+		return blk, nil
+
+	case bstore.ErrNotFound:
+		return s.cold.Get(cid)
+
+	default:
+		return nil, err
+	}
+}
+
+func (s *SplitStore) GetSize(cid cid.Cid) (int, error) {
+	size, err := s.hot.GetSize(cid)
+
+	switch err {
+	case nil:
+		return size, nil
+
+	case bstore.ErrNotFound:
+		return s.cold.GetSize(cid)
+
+	default:
+		return 0, err
+	}
+}
+
+func (s *SplitStore) Put(blk blocks.Block) error {
+	epoch := s.curTs.Height()
+	err := s.snoop.Put(blk.Cid(), epoch)
+	if err != nil {
+		return err
+	}
+
+	return s.hot.Put(blk)
+}
+
+func (s *SplitStore) PutMany(blks []blocks.Block) error {
+	err := s.hot.PutMany(blks)
+	if err != nil {
+		return err
+	}
+
+	epoch := s.curTs.Height()
+
+	batch := make([]cid.Cid, 0, len(blks))
+	for _, blk := range blks {
+		batch = append(batch, blk.Cid())
+	}
+
+	return s.snoop.PutBatch(batch, epoch)
+}
+
+func (s *SplitStore) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	chHot, err := s.hot.AllKeysChan(ctx)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	chCold, err := s.cold.AllKeysChan(ctx)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	ch := make(chan cid.Cid)
+	go func() {
+		defer cancel()
+		defer close(ch)
+
+		for _, in := range []<-chan cid.Cid{chHot, chCold} {
+			for cid := range in {
+				select {
+				case ch <- cid:
```

Comment:
This will be problematic. Either we open the second iterator when the first one finishes, or we open both and interleave the receives, thus consuming in parallel.

Reply:
why? They are separate stores, it shouldn't be a problem.

Reply:
Keeping an iterator open when you know you will not be making progress is not good.

Reply:
I just wanted to avoid writing the same code twice, but fine.

Reply:
there is a semantic problem here, we can't report the errors from getting the iterators if we move them to the goroutine.
I am inclined to leave it as is, I don't think it's a big deal if we hold an read iterator a little longer than necessary.

Reply:
Yeah, that's a good point. This interface has given us a lot more pain than gain.

Reply:
The problem is that it's not us who determines the speed of consumption, it's the caller. So if the caller slows down consumption during the processing of the hot set, the cold set iteration would've stayed open for too long unnecessarily. Depending on underlying database, the overhead will vary.

Reply:
Agreed, but otoh I don't think we are actually using this interface.
We can revisit if this becomes a problem.